win32ss/gdi/eng/pandisp.c

```diff
@@ -0,0 +1,484 @@
+    if (!pandev->surfObjScreen)
+    {
+        DPRINT1("Failed to lock underlying surface\n");
+        goto cleanup;
+    }
+
+    /* Create device surface */
+    pandev->hsurf = EngCreateDeviceSurface(NULL, pandev->szlDesktop, pandev->iBitmapFormat);
+    if (!pandev->hsurf)
+    {
+        DPRINT1("EngCreateDeviceSurface() failed\n");
+        goto cleanup;
+    }
+
+    if (!EngAssociateSurface(pandev->hsurf, pandev->hdev, HOOK_ALPHABLEND |
+                                                          HOOK_BITBLT |
+                                                          HOOK_COPYBITS |
+                                                          HOOK_GRADIENTFILL |
+                                                          HOOK_STROKEPATH |
+                                                          HOOK_SYNCHRONIZE |
+                                                          HOOK_TEXTOUT))
+
+    {
+        DPRINT1("EngAssociateSurface() failed\n");
+        goto cleanup;
+    }
+
+    /* Create shadow surface */
+    pandev->hsurfShadow = (HSURF)EngCreateBitmap(pandev->szlDesktop, pandev->szlDesktop.cx, pandev->iBitmapFormat, BMF_TOPDOWN, NULL);
+    if (!pandev->hsurfShadow)
+    {
+        DPRINT1("EngCreateBitmap() failed\n");
+        goto cleanup;
+    }
+
+    pandev->psoShadow = EngLockSurface(pandev->hsurfShadow);
+    if (!pandev->psoShadow)
+    {
+        DPRINT1("EngLockSurface() failed\n");
+        goto cleanup;
+    }
+
+    return pandev->hsurf;
+
+cleanup:
+    PanDisableSurface(dhpdev);
+    return NULL;
+}
+
+BOOL
+APIENTRY
+PanBitBlt(
+    _Inout_ SURFOBJ *psoTrg,
+    _In_opt_ SURFOBJ *psoSrc,
+    _In_opt_ SURFOBJ *psoMask,
+    _In_ CLIPOBJ *pco,
+    _In_opt_ XLATEOBJ *pxlo,
+    _In_ RECTL *prclTrg,
+    _In_opt_ POINTL *pptlSrc,
+    _In_opt_ POINTL *pptlMask,
+    _In_opt_ BRUSHOBJ *pbo,
+    _In_opt_ POINTL *pptlBrush,
+    _In_ ROP4 rop4)
+{
+    PPANDEV pandev;
+    BOOL res;
+
+    if (psoTrg->iType == STYPE_DEVICE)
+    {
+        pandev = (PPANDEV)psoTrg->dhpdev;
+        psoTrg = pandev->psoShadow;
+    }
+    if (psoSrc && psoSrc->iType == STYPE_DEVICE)
+    {
+        pandev = (PPANDEV)psoSrc->dhpdev;
+        psoSrc = pandev->psoShadow;
+    }
+
+    res = EngBitBlt(psoTrg, psoSrc, psoMask, pco, pxlo, prclTrg, pptlSrc, pptlMask, pbo, pptlBrush, rop4);
+
+    /* FIXME: PanSynchronize must be called periodically, as we have the GCAPS2_SYNCTIMER flag.
+     * That's not the case, so call PanSynchronize() manually */
+    if (res)
+        PanSynchronize(gPan, NULL);
+
+    return res;
+}
+
+BOOL
+APIENTRY
+PanCopyBits(
+    _Inout_ SURFOBJ *psoDest,
+    _In_opt_ SURFOBJ *psoSrc,
+    _In_ CLIPOBJ *pco,
+    _In_opt_ XLATEOBJ *pxlo,
+    _In_ RECTL *prclDest,
+    _In_opt_ POINTL *pptlSrc)
+{
+    /* Easy. Just call the more general function PanBitBlt */
+    return PanBitBlt(psoDest, psoSrc, NULL, pco, pxlo, prclDest, pptlSrc, NULL, NULL, NULL, ROP4_SRCCOPY);
+}
+
+BOOL
+APIENTRY
+PanStrokePath()
```

Comment:
Add maybe the actual parameters (in case the function gets called from function table) in order to avoid call stack problems? (Here and function below)